### PR TITLE
✨Add DeepCopy generation for additional types in builders

### LINKED ISSUE #6016: Add DeepCopy generation for additional types in builders
https://github.com/kubernetes-sigs/cluster-api/pull/5987 Adds new DeepCopy types to some of the builders used in our tests. These methods allow the same builder to be used to generate multiple test objects which should reduce some of the copy-pasting in test code.

Some of the types in the builders file do not have DeepCopy methods generated yet. This is because they contain a field that is `map[string]interface{}` which the [controller gen tool is not compatible with](https://github.com/kubernetes-sigs/controller-tools/blob/a410b23ff25739563e103e98af887adfd05feb98/pkg/deepcopy/traverse.go#L386-L393).

These types are currently excluded using the `kubebuilder:object:generate=false` tag. We should restructure these builder objects to allow them to use code generation by changing their `map[string]interface{}` to a type that controller-gen will be able to generate DeepCopy methods for e.g. apiextensionsv1.JSON{} and generate DeepCopy methods for those types.

c/f https://github.com/kubernetes-sigs/cluster-api/pull/5987#discussion_r794413897

/area testing
/good-first-issue

### PR DESCRIPTION
**Pull request by sayantani11**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #6016

### CODE REVIEW

**Comment by sayantani11** [comment]
cc @killianmuldoon

File: internal/test/builder/builders.go [hunk 1]
```diff
@@ -419,60 +420,58 @@ func InfrastructureMachineTemplate(namespace, name string) *InfrastructureMachin
...
+		obj,
 	}
+
 }
 
 // WithSpecFields will add fields of any type to the object spec. It takes an argument, fields, which is of the form path: object.
 func (b *BootstrapTemplateBuilder) WithSpecFields(fields map[string]interface{}) *BootstrapTemplateBuilder {
-	b.specFields = fields
+	for field, value := range fields {
+		unstructured.SetNestedField(i.obj.UnstructuredContent(), value, field)
```
Comment (killianmuldoon) [inline internal/test/builder/builders.go:462]: Could you change this and rerun the tests? 
```suggestion
		unstructured.SetNestedField(b.obj.UnstructuredContent(), value, field)
```

**Review by killianmuldoon — APPROVED** [review]

**Comment by killianmuldoon** [comment]
/retitle ✨ [WIP]  Add DeepCopy generation for additional types in builders #6302

**Comment by killianmuldoon** [comment]
/retitle ✨ [WIP] Add DeepCopy generation for additional types in builders

File: internal/test/builder/builders.go (hunk 1 above)
Comment (sayantani11) [reply internal/test/builder/builders.go:462]: Yeah sure!!! That was a mistake

**Comment by sayantani11** [comment]
Hey @killianmuldoon I was caught up with exams and stuff will push the PR within today!!! Sorry for the delay

**Comment by killianmuldoon** [comment]
@sayantani11 I see the current version of the code isn't building - are you having trouble resolving the errors?

**Comment by killianmuldoon** [comment]
@sayantani11 I opened a PR at https://github.com/sayantani11/cluster-api/pull/1 focusing on make one single type deepcopy-able. 

Could try to do the same for one other type - e.g. BootstrapTemplateBuilder - so that we can make the changes incrementally to get a pattern of how to solve any obstacles you find along the way.

**Comment by sayantani11** [comment]
> @sayantani11 I opened a PR at [sayantani11#1](https://github.com/sayantani11/cluster-api/pull/1) focusing on make one single type deepcopy-able.
> 
> Could try to do the same for one other type - e.g. BootstrapTemplateBuilder - so that we can make the changes incrementally to get a pattern of how to solve any obstacles you find along the way.

Okay will update soon!!

**Comment by sayantani11** [comment]
Hey!!!
Sorry for the delay! I actually have my semester exam ongoing 😓 so not
really able to look at it right now! I'll do this weekend. Sorry about that.

**Comment by killianmuldoon** [comment]
No worries @sayantani11 - if you don't have time we can close this for now and you can pick something up when you've got a bit more free time. 

Best of luck with the exams!

**Comment by killianmuldoon** [comment]
@sayantani11 I'm going to close this for now, but if you find time to pick it up again feel free to. If you think you won't have time (completely understandable :slightly_smiling_face: ) then you can unassign yourself from the issue.

**Pull request closed without merging**